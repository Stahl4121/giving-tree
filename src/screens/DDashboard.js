import React, { useEffect } from 'react';
import SmallGrantCard from '../components/SmallGrantCard.js';
import Grid from '@material-ui/core/Grid';
import Container from '@material-ui/core/Container';
import { useCollection } from 'react-firebase-hooks/firestore';
import firebase from '../firebase.js';
import Search from '../components/Search';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  container: {
    paddingBottom: theme.spacing(2),
  },
}));

export default function DDashboard() {
  // List of small grant cards
  const [grants, setGrants] = React.useState([]);

  // Initialize database and specific grant in database
  const db = firebase.firestore();

  const [snapshot, loading, error] = useCollection(db.collection('grants'));

  useEffect(() => {
    var newGrants = [];
    if (!loading && !error) {
      snapshot.forEach(function (doc) {
        newGrants.push(
          <SmallGrantCard
            id={doc.id}
            title={doc.data().title}
            cfName={doc.data().cf_name}
            nonprofitName={doc.data().nonprofit_name}
            goalAmt={doc.data().goal_amt}
            moneyRaised={doc.data().money_raised}
            img={doc.data().images[0] || 'GivingTree.png'} />
        );
      });
      setGrants(newGrants);
    }
  }, [snapshot, error, loading]);
  /*
  useEffect(() => {
    var useless = 0;
  }, [grants]);
  */
  function searchCallback(childData) {
    var newGrants = [];
    childData.forEach((grant) => {
      newGrants.push(
        <SmallGrantCard
          id={grant.id}
          title={grant.title}
          cfName={grant.cf_name}
          nonprofitName={grant.nonprofit_name}
          goalAmt={grant.goal_amt}
          moneyRaised={grant.money_raised}
          img={grant.images[0] || 'GivingTree.png'} />
      );
    });
    setGrants(newGrants);
  }

  const classes = useStyles();

  return (
    <Container maxWidth='md' className={classes.container}>
      <Search parentCallback={searchCallback} />
      <Grid container spacing={2} >
        {grants}
      </Grid>
    </Container>
  );
}
