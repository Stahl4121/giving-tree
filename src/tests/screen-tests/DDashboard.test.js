import '@testing-library/jest-dom'
import React from 'react'
import { shallow, mount } from 'enzyme'
import { render, screen } from '@testing-library/react'
import DDashboard from '../../screens/DDashboard'
import FirestoreMock from '../firestore.mock'
import * as FirebaseHooks from 'react-firebase-hooks/firestore';

const mockCfDocs = [{
  data: () => {
    return {
      dist: -1,
      id: "ASDF1234asdf",
      title: "Grant Title",
      cfName: "Calvis Community Foundation",
      nonprofitName: "NonProf Name",
      goalAmt: 300000000000,
      moneyRaised: 0,
      img: 'GivingTree.png',
      nonprofitId: "4a1MMPzzdmREry9CUV6T",
      address: "1911 C Street NE, Washington D.C. 20002",
      lat: 40,
      long: -70,
      datePosted: 'March 09, 2020 at 9:09:34 AM UTC-4',
      dateDeadline: 'March 09, 2020 at 9:09:34 AM UTC-4',
      desc: 'gimme that cash',
      status: "deleted",
      tags: [],
    }
  }
},
{
  data: () => {
    return {
      dist: -1,
      id: "ASDF1234asdf",
      title: "Grant Title",
      cfName: "Calvis Community Foundation",
      nonprofitName: "NonProf Name",
      goalAmt: 300000000000,
      moneyRaised: 0,
      img: 'GivingTree.png',
      nonprofitId: "4a1MMPzzdmREry9CUV6T",
      address: "1911 C Street NE, Washington D.C. 20002",
      lat: 40,
      long: -70,
      datePosted: 'March 09, 2020 at 9:09:34 AM UTC-4',
      dateDeadline: 'March 09, 2020 at 9:09:34 AM UTC-4',
      desc: 'gimme that cash',
      status: "deleted",
      tags: [],
    }
  }
}
]

const mockFirestore = new FirestoreMock();

jest.mock('firebase/app', () => ({
  __esModule: true,
  initializeApp: () => { },
  apps: {
    length: true,
  },
  app: () => {
    return {
      auth: () => { },
      firestore: () => { return mockFirestore },
    }
  },
}));

describe('DDashboard', () => {

  test('DDashboard Basics', () => {
    jest.mock('react-router', () => ({
      useParams: jest.fn().mockReturnValue({ id: '123' }),
    }));
    
    render(<DDashboard />);
    expect(screen).toMatchSnapshot();
  });
  
  describe('DDashboard with mocked data', () => {
    it('uses the search callback correctly', () => {
  
    })
  })
  
  describe('uses firestore', () => {
    beforeEach(() => {
      const forEachMock = jest.fn(() => { return mockCfDocs });
      const snapshot = { forEach: forEachMock };
      FirebaseHooks.useCollection = () => { return [snapshot, false, false] };
      mockFirestore.reset();
    });
  
    it('call grants snapshot on render', () => {
      const wrapper = mount(<DDashboard />);
      wrapper.update();
      expect(mockFirestore.mockCollection).toBeCalledWith('grants');
      expect(mockFirestore.mockCollection).toHaveBeenCalledTimes(2);
    });
  });

})
