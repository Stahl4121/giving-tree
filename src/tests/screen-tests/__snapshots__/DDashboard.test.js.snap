// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DDashboard DDashboard Basics 1`] = `ShallowWrapper {}`;
